components/raftstore/src/store/fsm/apply.rs

```diff
@@ -1830,19 +1843,101 @@ where
             {
                 peer.set_id(*peer_id);
             }
-            write_peer_state(kv_wb_mut, &new_region, PeerState::Normal, None)
+            new_regions_map.insert(
+                new_region.get_id(),
+                (
+                    util::find_peer(&new_region, ctx.store_id).unwrap().get_id(),
+                    None,
+                ),
+            );
+            regions.push(new_region);
+        }
+
+        if right_derive {
+            derived.set_start_key(keys.pop_front().unwrap());
+            regions.push(derived.clone());
+        }
+
+        let mut replace_regions = HashSet::default();
+        {
+            let mut pending_create_peers = ctx.pending_create_peers.lock().unwrap();
+            for (region_id, (peer_id, reason)) in new_regions_map.iter_mut() {
+                match pending_create_peers.get_mut(region_id) {
+                    Some(status) => {
+                        if *status != (*peer_id, false) {
+                            *reason = Some(format!("status {:?} is not expected", status));
+                        } else {
+                            replace_regions.insert(*region_id);
+                            pending_create_peers.insert(*region_id, (*peer_id, true));
```

Comment:
Is it safe to replace? What if the split range is not overlap with the replicated one?

Reply:
Yes, it's safe. This peer id is the same as the previous replicated one. This region can't merge the other region because one peer does not exist which violates the guarantee that all target peer must exist during merging. If this really happens, the panic will happen at https://github.com/tikv/tikv/blob/4a75902f266fbbc064f0c19a2a681cfe66511bc3/components/raftstore/src/store/fsm/peer.rs#L1896-L1910

Reply:
What about a new peer created by replication?